### PR TITLE
Full `#[derive(SolidityError)]`

### PR DESCRIPTION
**Pull request by rachel-bousfield**
This PR completes the `#[derive(SolidityError)]` macro. Now users can annotate error enums like
```rs
sol! {
    error InsufficientBalance(address from, uint256 have, uint256 want);
    error InsufficientAllowance(address owner, address spender, uint256 have, uint256 want);
}

#[derive(SolidityError)]
pub enum Erc20Error {
    InsufficientBalance(InsufficientBalance),
    InsufficientAllowance(InsufficientAllowance),
}

#[external]
impl Contract {
    pub fn fallible_method() -> Result<(), Erc20Error> {
        // code that might revert
    }
}
```

And automatically export ABIs like
```sh
cargo stylus export-abi
```

```solidity
interface IContract {
    function fallibleMethod() external;

    error InsufficientBalance(address, uint256, uint256);

    error InsufficientAllowance(address, address, uint256, uint256);
}
```

This was made possible by
- Adding a new `InnerTypes` trait for collecting errors and structs (we'll reuse this mechanism for exporting struct definitions once they're allowed in method signatures).
- Gracefully handling errors to improve DevX when annotation fails
- Adding docstrings and other minor improvements

### CODE REVIEW

**Review by mikebenfield — COMMENTED** [review]
Just a couple comments.

File: stylus-proc/src/methods/error.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    let mut match_arms = quote!();
+    let mut errors = vec![];
+    for variant in input.variants {
+        let variant_name = variant.ident;
+        let error = match variant.fields {
+            Fields::Unnamed(e) if variant.fields.len() == 1 => e.unnamed.first().unwrap().clone(),
+            _ => error!(variant.fields, "Variant not a 1-tuple"),
+        };
+        match_arms.extend(quote! {
+            #name::#variant_name(e) => e.encode(),
```
[the remark this answers is not in the document]
Comment (rachel-bousfield) [reply stylus-proc/src/methods/error.rs]: Makes sense, updated :)

File: stylus-sdk/src/abi/internal.rs [hunk 2]
```diff
@@ -1,19 +1,35 @@
...
 use alloc::{vec, vec::Vec};
 use alloy_primitives::U256;
 use alloy_sol_types::SolType;
 use core::fmt;
 
-#[inline(always)]
-pub fn encode_return_type<T: AbiType>(x: T) -> Vec<u8> {
-    // coerce types into a tuple of at least 1 element
-    <(T,) as AbiType>::SolType::encode(&(x,))
+pub trait EncodableReturnType {
```
[the remark this answers is not in the document]
Comment (rachel-bousfield) [reply stylus-sdk/src/abi/internal.rs:13]: This was from #103 so that you could use both `Result<T>` and `T` types in method signatures

**Review by mikebenfield — APPROVED** [review]
LGTM

**Pull request merged**